File: src/dp/5.longest-palindromic-substring/5.longest-palindromic-substring.ts
/**
 * Time complexity: O(n ^ 2)
 * Space complexity: O(n ^ 2)
 *
 * @param s
 */
function longestPalindrome(s: string): string {
  return longestPalindromeVersion2(s);
}

/**
 * Time complexity: O(n ^ 2)
 * Space complexity: O(n ^ 2)
 *
 * @param s
 */
function longestPalindromeVersion1(s: string): string {
  let left = 0, right = 0;
  const arr = Array.from(s),
        n   = arr.length,
        dp  = Array.from(Array(n), () => Array(n).fill(false));

  for (let i = n - 2; i >= 0; i--) {
    for (let j = i; j < n; j++) {
      if (j - i <= 2)
        dp[i][j] = arr[i] === arr[j];
      else
        dp[i][j] = dp[i + 1][j - 1] && arr[i] === arr[j];

      if (dp[i][j] && j - i > right - left) {
        left = i;
        right = j;
      }
    }
  }

  return s.substring(left, right + 1);
}

longestPalindrome("babad");

/**
 * Time complexity: O(n ^ 2)
 * Space complexity: O(n ^ 2)
 *
 * @param s
 */
function longestPalindromeVersion2(s: string): string {
  let left = 0, right = 0;
  const arr = Array.from(s),
        n   = arr.length

  for (let i = 0; i < n -1; i++) {
    const len = Math.max(expandAroundCenter(i, i), expandAroundCenter(i, i + 1))

    if (len > right - left + 1) {
      left = i - ((len - 1) >> 1)
      right = i + (len >> 1)
    }
  }

  return s.substring(left, right + 1);

  function expandAroundCenter(left: number, right: number): number {
    while (left >= 0 && right < n && arr[left] === arr[right]) {
      left--
      right++
    }
    return right - left - 1
  }
}

longestPalindrome("babad");

File: src/dp/5.longest-palindromic-substring/5.longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        return longestPalindromeVersion2(s)


def longestPalindromeVersion1(s: str) -> str:
    """
    Time complexity: O(n^2)
    Space complexity: O(1)
    :param s:
    :return:
    """
    left, right, n = 0, 0, len(s)

    def expandAroundCenter(l: int, r: int) -> int:
        while l >= 0 and r < n and s[l] == s[r]:
            l -= 1
            r += 1
        return r - l - 1

    for i in range(n - 1):
        length = max(expandAroundCenter(i, i), expandAroundCenter(i, i + 1))
        if length > right - left + 1:
            left = i - (length - 1) // 2
            right = i + length // 2

    return s[left:right + 1]


def longestPalindromeVersion2(s: str) -> str:
    """
    Manacher algorithm
    Time complexity: O(n)
    Space complexity: O(n)
    :param s:
    :return:
    """
    t = "#" + "#".join(s) + "#"

    RL = [0] * len(t)
    max_right = 0
    pos = 0

    for i in range(len(t)):
        if i < max_right:
            RL[i] = min(RL[2 * pos - i], max_right - i)
        else:
            RL[i] = 1

        # 尝试扩展，注意处理边界
        while i - RL[i] >= 0 and i + RL[i] < len(t) and t[i - RL[i]] == t[i + RL[i]]:
            RL[i] += 1
        # 更新 MaxRight, pos
        if RL[i] + i - 1 > max_right:
            max_right = RL[i] + i - 1
            pos = i
    max_len, center_index = max((n-1, i) for i, n in enumerate(RL))
    return s[(center_index//2 - max_len//2):  (center_index//2 - max_len//2) + max_len ]

x = Solution()
# x.longestPalindrome("babad")
# x.longestPalindrome("cbbd")
x.longestPalindrome("a")